### ISSUE TITLE
JS Minification

### ISSUE THREAD

**Issue by Lux-Delux** [jacobwb/hashover-next#107]
I know this isn't easy to do with the current implementation due to various possible setups.

Perhaps we can brainstorm on possible solutions.

For example my loaded hashover.php? ... can be minified by at least 21%.

Perhaps make minification 1 time after the settings are in place and then load the minified resource from there on forward?

Using something like this perhaps https://github.com/tedious/JShrink not sure about licensing etc.

**Comment by jacobwb** [comment]
> I know this isn't easy to do with the current implementation due to various possible setups.

I will be rewriting most of the JavaScript once I'm done removing the inline HTML and merging the JavaScript and PHP modes. Among the things that will change are...

1. Dynamic information the JavaScript expects will be stated before the majority of the JavaScript runs. Meaning work for things like settings, what forms will be displayed, etc, will be done first and the rest of the JavaScript will be static/unchanging. This will allow for easier minification.

2. All HTML is dynamically generated now, meaning that multi-line and single-line output of the resultant HTML code can now be configurable. Identical to PHP's JSON "Pretty Print."

In short, minification will soon be much easier to implement.

> For example my loaded hashover.php? ... can be minified by at least 21%.

I would assume a good portion of that is just the JSON.

For example:

```json
"1": {
	"avatar": "\/hashover\/scripts\/avatars.php?format=png&amp;size=45&amp;email=818c7bfdd8ed3dd7d96079c347dd23a4",
	"permalink": "c1",
	"name": "Jacob Barkdull",
	"website": "http:\/\/tildehash.com\/?page=author",
	"date": "45 years ago",
	"sort_date": 0,
	"likes": "0",
	"comment": "Comment..."
},
"2": {
	"avatar": "\/hashover\/scripts\/avatars.php?format=png&amp;size=45&amp;email=818c7bfdd8ed3dd7d96079c347dd23a4",
	"permalink": "c2",
	"name": "Jacob Barkdull",
	"website": "http:\/\/tildehash.com\/?page=author",
	"date": "45 years ago",
	"sort_date": 0,
	"likes": "0",
	"comment": "Second comment..."
}
```

When minified is 19.74% smaller. Considering that after 100 comments or so the JSON data is the majority of the JavaScript code, disabling `JSON_PRETTY_PRINT` may help a lot by itself.

> Perhaps make minification 1 time after the settings are in place and then load the minified resource from there on forward?

I don't know exactly what the best way to do it would be.

For example, minified JavaScript in general has the potential of obfuscating the code. In fact, in most "minification" libraries obfuscation is done as part of the minification process.

After all:

```javascript
(function(){function a(a){alert(a)}a('Hi!')})();
```

Is much smaller than:

```javascript
(function(){function show_message(message){alert(message)}show_message('Hi!')})();
```

The difference is the first is hard to follow and may not do what you think it does, which allows people (developers) to sneak in malicious features. While the second can be read, followed, is clear about what it does, and can be unminified back into its original source code.

The first is also harder to do with dynamically generated JavaScript than the second.

Personally, minified JavaScript is something I hate, but I do see the performance and bandwidth benefits of it. So my only concern in the end is the user's ability to obtain the original unminified source code. So, I think the minification solution needs to be based on the original unminified JavaScript in such a way that the minified version can't exist without the unminified version.

Then the unminified JavaScript would have a comment pointing users to the executing minified code in it's original form. In fact, that might be required by the AGPL. I'm not sure.

**Comment by Lux-Delux** [comment]
Nice, I'm looking forward to the updates.

I basically have hashover optimized when it comes to page render with only the unminified code left as a bandwidth "problem".

As a first step, perhaps just collapsing whitespace on the output would prove beneficial, and as you mentioned `JSON_PRETTY_PRINT`.

I wouldn't have even noticed this but it was one of the last "issues" I had with pagespeed insights

**Comment by Lux-Delux** [comment]
Not sure whether this will be the issue with the upcoming changes but anyway the "problem" actually lies with the addtional hashover.php?=url script being added to head as an additional script.

Could you explain a bit how this works and why is it necessary? Is it not possible for all the php processing to be done just by calling hashover.js (without creating another script call)?

If it was the other way around then stuff like mod/ngx pagespeed would be able to automagically compress the output.

**Comment by jacobwb** [comment]
Soon HashOver will no longer add anything to the `<head>` element, everything will go in `<body>` in the future, but that's probably not the problem if the additional script is the issue.

Technically, yes. The `hashover.js` file is just for convenience. Basically, there is no way for PHP to obtain the page title, nor the URL very reliably really. And so `hashover.js` is used to get the page URL and title and pass those to the HashOver PHP as URL parameters.

Many people don't have a way to easily pass around page information, so `hashover.js` is there to do that work for them automatically. There's no need to use it if you'll get better performance without it.

You could instead point the HashOver script tag directly to:

    /hashover/hashover.php?url=<URL>&title=<TITLE>

And that would eliminate the need for `hashover.js`

**Comment by Lux-Delux** [comment]
Interesting, I'll try that out. Let's see if pagespeed recognizes that

**Comment by Lux-Delux** [comment]
Turns out it's the same either way so I kept using hashover.js

**Comment by jacobwb** [comment]
> Using something like this perhaps https://github.com/tedious/JShrink not sure about licensing etc.

After some testing, I've decided JShrink is inefficient. My tests were of 1,000 comments parsed a thousand times, at random times. Below are the average statistics.

Base (normal unminified JavaScript)
---
<table>
<tr>
<td>JavaScript Size</td>
<td>810,346 bytes</td>
</tr>
<tr>
<td></td>
<td></td>
</tr>
<tr>
<td>Execution Time</td>
<td>0.25443 Seconds</td>
</tr>
<tr>
<td>Script Peak Memory Usage</td>
<td>5.35Mb</td>
</tr>
<tr>
<td>System Peak Memory Usage</td>
<td>7.5Mb</td>
</tr>
</table>

JShrink
---
<table>
<tr>
<td>JavaScript Size</td>
<td>752,541 bytes</td>
</tr>
<tr>
<td>Size Reduction</td>
<td>7.13%</td>
</tr>
<tr>
<td></td>
<td></td>
</tr>
<tr>
<td>Execution Time</td>
<td>0.34076 Seconds</td>
</tr>
<tr>
<td>Execution Time Difference</td>
<td>+0.08633 Seconds &#8595;</td>
</tr>
<tr>
<td>Script Peak Memory Usage</td>
<td>4.42Mb &#8593;</td>
</tr>
<tr>
<td>System Peak Memory Usage</td>
<td>6Mb &#8593;</td>
</tr>
</table>

_It's worth noting that JShrink doesn't fully minify the JavaScript. It leaves many newlines, and doesn't remove unnecessary semicolons._

Custom Minifier (so far)
---
<table>
<tr>
<td>JavaScript Size</td>
<td>751,581 bytes</td>
</tr>
<tr>
<td>Size Reduction</td>
<td>7.25%</td>
</tr>
<tr>
<td></td>
<td></td>
</tr>
<tr>
<td>Execution Time</td>
<td>0.24631 Seconds</td>
</tr>
<tr>
<td>Execution Time Difference</td>
<td>-0.00812 Seconds &#8593;</td>
</tr>
<tr>
<td>Script Peak Memory Usage</td>
<td>4.42Mb &#8593;</td>
</tr>
<tr>
<td>System Peak Memory Usage</td>
<td>6Mb &#8593;</td>
</tr>
</table>

My custom minifier is the clear winner.

Though, I'm seeing some odd things happening, like, minified code doesn't have spaces between things such as array items and function parameters, but there's a bug where something like...

```javascript
var products = ['milk', 'eggs', 'plutonium'];
```

Would lose the spaces, but so would...

```javascript
var message = 'Welcome, user!';
```

This is also happening for things like parentheses, like "Showing 27 Comments (31 counting replies)" is "Showing 27 Comments(31 counting replies)".

This is probably because I'm being really lazy with the minification parser. I'm putting emphasis on the speed of the minification process, not the "smartness" of it.

I'll probably just move the locale strings and such out of the actual JavaScript, and "inject" them after the JavaScript is minified. That should work. Those things are shared by both the JavaScript and PHP mode so they should probably be separated anyway.

**Comment by Lux-Delux** [comment]
Nice, didn't expect you already working on this :)

Glad to see your solution performs better. Actually it makes me really happy as I'm a obsessed with performance hehe. Looking forward to testing this
